daemon: Allow releasing builder while waiting for proxy ACKs

jrajahalme opened the pull request:
Sidecar proxy ACKs can take a long time due to container
initialization. Allow other Endpoints to build while another Endpoint
is waiting for proxy completions.

Build queue is implemented with a semaphore to simplify the implementation. Queueing for the build new makes the calling go-routine wait when Acquiring the "build permission". A function for releasing the permission is returned if a permission is granted, nil otherwise. The permission is released after the endpoint bpf compilation and before waiting for the proxy completion so that endpoint builds are not blocked when other endpoints are waiting for proxy completions. This allows the overall build timeout to be increased so that endpoint builds need not fail because of slow container initialization.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6211)
<!-- Reviewable:end -->

jrajahalme committed:
endpoint: Make regenaration timeout greater than ExecTimeout

ExecTimeout is used while the overall regeneration timeout is already
ticking. Before it was possible that BPF compilation took a long time,
but not exeeding `ExecTimeout`, but the overall regeneration would
still fail right after due to proxyWaitGroup having timed out during
BPF compilation.

This is now safer than before as other endpoints can start building
while other endpoints wait for proxy completion.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Eliminate ExecTimeout, ctx.

Use the overall regeneration context instead.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

raybejjani commented on daemon/daemon.go:
```diff
@@ -214,65 +215,50 @@ func (d *Daemon) RemoveNetworkPolicy(e *endpoint.Endpoint) {
+	return func() {
+		if !done {
+			d.buildEndpointSem.Release(1)
+			done = true
```
I'm still reading through but this may be racey if the returned function can be called from multiple places. If true, it might work to just:
```
	select {
	case <-done:
		return
	default:
		close(done)
		d.buildEndpointSem.Release(1)
	}
```
but I'm not sure if the default and `close(done)` will race as well :/

jrajahalme replied:
Changed to use `sync.Once` as suggested by @aanm

raybejjani commented on daemon/daemon.go:
```diff
@@ -214,65 +215,50 @@ func (d *Daemon) RemoveNetworkPolicy(e *endpoint.Endpoint) {
-// RemoveFromEndpointQueue removes the endpoint from the queue.
-func (d *Daemon) RemoveFromEndpointQueue(epID uint64) {
+	// Acquire build permit
+	err := d.buildEndpointSem.Acquire(ctx, 1)
```
I was going to suggest moving this to the start of the function to avoid unlocking and locking ;) Then I realised why we do this. Expanding the comment to explicitly say how this semAcquire relates to `d.uniqueID`, and that while an endpoint is waiting to get the semaphore we don't want to queue it again, but once it can proceed (i.e. build) we are ok re-queuing.

jrajahalme replied:
I updated the comments all around.

raybejjani commented on daemon/daemon.go:
```diff
@@ -214,65 +215,50 @@ func (d *Daemon) RemoveNetworkPolicy(e *endpoint.Endpoint) {
-			}
-		}()
+// RemoveFromEndpointQueue removes the endpoint from the queue.
+func (d *Daemon) RemoveFromEndpointQueue(epID uint64) {
```
I can't find a spot where we use this function but this cancels the context so it's more of a "cancel the build" request, right? If so, it might be worth mentioning that in the comment above

raybejjani replied:
Oh, perhaps this needs to be called when an endpoint is removed from `endpointmanager`. A queued build probably shouldn't proceed in that case (but this might be handled in some other way... I don't think it's handled in the older code here)

jrajahalme replied:
This is called from `endpoint.LeaveLocked()` and will cancel waiting for a build permit. This call will not cancel a build already in progress, but no change in this regard.

raybejjani commented on pkg/endpoint/policy.go:
```diff
@@ -842,13 +841,12 @@ func (e *Endpoint) Regenerate(owner Owner, context *RegenerationContext) <-chan
-
+			context.DoneFunc = doneFunc
 			err := e.regenerate(owner, context)
+			doneFunc() // in case not called already
```
Won't this double release the semaphore if `doneFunc` was already called?

raybejjani replied:
oh, no, the done check inside doneFunc guards against that. 👍

raybejjani commented on pkg/endpoint/policy.go:
```diff
@@ -812,23 +816,18 @@ func (e *Endpoint) regenerate(owner Owner, context *RegenerationContext) (retErr
 			// make sure we don't block
 			select {
-			case req.ExternalDone <- buildSuccess:
+			case done <- buildSuccess:
```
Is this supposed to block the return of `Regenerate` or only handle the case where nothing is listening on the other end of this channel? I suspect the latter and, if I'm correct, we could make `done` a buffered channel and only defer the close on it. I think it works in all 3 cases:
- no one reads `done`. nothing happens either way
- the reader reads done after we write to it. This means the new scheme allows the caller of Regenerate to continue (this may break something, who knows)
- the reader is already waiting on `done`, nothing changes.

aanm replied:
@raybejjani 

>no one reads done. nothing happens either way

~go routines are leaked~

Spoke offline, I misread what he wrote. Yeah a buffered channel would be ideal here.

jrajahalme replied:
Changed to a buffered channel. Still need to defer both writing to the channel and closing it.

aanm commented on daemon/daemon.go:
```diff
@@ -214,65 +215,50 @@ func (d *Daemon) RemoveNetworkPolicy(e *endpoint.Endpoint) {
+	// Caller calls the returned function when the heavy lifting of the build is done.
+	done := false
+	return func() {
+		if !done {
```
`sync.Once` would do what you are trying to achieve here and it would protect you against parallel executions of this returned `func()`

jrajahalme replied:
Done! The done func was called only from one goroutine, but it's better to not race if that ever changes. Thanks for the suggestion!

aanm commented on daemon/state_test.go:
```diff
@@ -97,12 +97,24 @@ func (ds *DaemonSuite) generateEPs(baseDir string, epsWanted []*e.Endpoint, epsM
+
+	ds.OnQueueEndpointBuild = func(epID uint64) func() {
+		builders++
+		done := false
```
Same here for the `sync.Once` comment

jrajahalme replied:
Done.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -56,7 +56,10 @@ const (
+	// Must be greater than `ExecTimeout` as a context with this timeout
+	// is started before using `ExecTimeout` and still used after `ExecTimeout`
+	// is already canceled.
+	EndpointGenerationTimeout = 330 * time.Second
```
This is a tough one, if something takes more than 330 seconds to build it's just straight wrong. We can't expect that an endpoint took more than 20 seconds to enforce a policy. I know there are hardware constraints but we should get reasonable here in order to enforce a maximum acceptable value.

jrajahalme replied:
@aanm See the note https://github.com/cilium/cilium/pull/6211#issuecomment-437300178 above. It is quite common that the sidecar proxy container within an Istio pod may take 2-3 minutes to start up. If we time out before the sidecar proxy gets and acks the L7 network policy the endpoint ends up in `not-ready` state, which may linger indefinitely.

jrajahalme replied:
The point of this PR is to make the longer timeouts less harmful, by allowing other endpoints to regenerate as soon as the resource intensive parts of regeneration are done.

We already do return from the API call on the initial build without waiting for proxy completions. If I recall correctly the purpose of that is to allow the pod initialization to continue after L3/4 networking is up and running (@ianvernon correct if I'm wrong here!).

jrajahalme replied:
We also have seen actual bpf compilations take longer than a minute in the past on very resource constrained systems. This has improved a lot (and is still improving) for the typical case, but IMO the worst case scenario still exists?

Either way, I'd leave fine tuning the timeout for a separate PR.

jrajahalme committed:
daemon: Allow releasing builder while waiting for proxy ACKs

Sidecar proxy ACKs can take a long time due to container
initialization. Allow other Endpoints to build while another Endpoint
is waiting for proxy completions.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Use sync.Once, rewamp comments.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
